### PR TITLE
optimize pyparsing import time by deferring regex compile

### LINKED ISSUE #362: 3x import-time performance regression between 2.x and 3.x
👋 I'm a bit down a rabbit hole so I'll explain my journey and how I got here

I was taking a look at the newest version of pip today and I noticed it was visibly slower than before.  a pretty noticeable half second pause on startup (450-500ms)

I did a little bit of profiling and noticed that almost half of that time (minus the base interpreter startup) was spent importing `pyparsing` and it felt like it was slower than before

I poked around a little bit and noticed that the performance regressed pretty significantly between 2.4.7 and 3.0.0

here's a comparison of some of the import times there (note that these are ~best of 5 which isn't necessarily a super useful representation of speed -- but it's enough to show significance here)

## base interpreter startup

(essentially `import site`) -- I'm using python 3.8.10 here on ubuntu 20.04

```console
$ time python3 -c ''

real	0m0.020s
user	0m0.020s
sys	0m0.000s
```

## 2.4.7

```console
$ time python3 -c 'import pyparsing'

real	0m0.047s
user	0m0.043s
sys	0m0.004s
```

## 3.0.0

```console
$ time python3 -c 'import pyparsing'

real	0m0.169s
user	0m0.168s
sys	0m0.000s
```

and just to confirm it's still present on the latest version, this is 938f59d183ca790e41a1d995aa3d3e995df2fb19

```console
$ time python3 -c 'import pyparsing'

real	0m0.169s
user	0m0.169s
sys	0m0.000s
```

## profiling

I did a bit of profiling, it looks like it spends ~75% of the time compiling regular expressions -- most of it coming from `pyparsing.core:Regex.__init__` -- I'll attach a svg of the profile below

<summary>profile svg:
<details>

log

</details>
</summary>

## regressions (?)

<summary>I did some bisection using this script:
<details>

```python
import argparse
import subprocess
import sys
import time


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('--best-of', type=int, default=5)
    parser.add_argument('--cutoff', type=float, default=.1)
    args = parser.parse_args()

    best = None

    for i in range(args.best_of):
        t0 = time.monotonic()
        subprocess.check_call((sys.executable, '-c', 'import pyparsing'))
        t1 = time.monotonic()

        duration = t1 - t0
        if best is None or duration < best:
            best = duration

    print(f'best of {args.best_of}: {best:.3f}')
    return best >= args.cutoff


if __name__ == '__main__':
    raise SystemExit(main())
```

</details>
</summary>

the nice thing about this script is it lets me set thresholds and find each regression here

here's some thresholds as well as their commits:

- 90ms threshold, 98ms best of 5: e2fb9f25431544b3c783e13e7fffc0e17bcf9fc8
- 120ms threshold, 163ms best of 5: aab37b6838b2f5be43c137ca17a40d5f22d3279f

these seem to be the two most impactful changes to startup time

## ideas

optimizing this is potentially a little annoying, but I think the best thing that could be done here is to lazily compile those regular expressions -- perhaps by moving the regex object to a property?  or lazily constructing the `Regex` instance ? or if python3.7+ is targetted it could use module-level `__getattr__` to get some further wins

### PR DESCRIPTION
**Pull request by asottile**
resolves #362

using the same script in that PR:

```console
$ python3 -m prof --best-of 10
0.066
```

### CODE REVIEW

**Comment by ptmcg** [comment]
Looks great, thanks!

File: pyparsing/core.py [hunk 1]
```diff
@@ -2947,18 +2951,36 @@ def __init__(
...
         self.asMatch = asMatch
         if self.asGroupList:
             self.parseImpl = self.parseImplAsGroupList
         if self.asMatch:
             self.parseImpl = self.parseImplAsMatch
 
+
+    @cached_property
+    def re(self):
+        if self._re:
```
Comment (RonnyPfannschmidt) [inline pyparsing/core.py:2966]: Wouldn't it be possible to always assign to re and thus remove the extra attribute for a given pattern

Comment (asottile) [reply pyparsing/core.py:2966]: no, see `__init__`

Comment (RonnyPfannschmidt) [reply pyparsing/core.py:2966]: @asottile my understanding is that a cached property can be assigned a value using the same attribute name, as its using the __dict__ as cache

that way the assigment should suffice

Comment (asottile) [reply pyparsing/core.py:2966]: yes, but `self._re` is used separately

**Pull request merged**